### PR TITLE
[draft] [Coop] InterlockedExchange(object), InterlockedCompareExchange(object).

### PR DESCRIPTION
**Pull request by jaykrell**
- Usually intrinsified.
 - Always?
 - Avoid coop handles by manually always using ref to object.
 - For efficiency on input, and efficiency wrt MonoError.
 - For correctness on output (due to aliasing in the wrappers).
 - Unclear on location.

### CODE REVIEW

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by vargaz** [comment]
@monojenkins build deb with monolite

File: mcs/class/corlib/System.Threading/Interlocked.cs [hunk 1]
```diff
@@ -50,7 +50,27 @@ public static partial class Interlocked
...
+			// location1's treatment is unclear. But note that passing it by handle would be incorrect,
+			// as it would use a local alias, which the coop marshaling does, to avoid the unclarity here,
+			// that of a ref being to a managed frame vs. a native frame. Perhaps that could be revisited.
+			//
+			// So there a hole here, that of calling this function with location1 being in a native frame.
+			// Usually it will be to a field, static or not, and not even to managed stack.
+			//
+			// This is usually intrinsified. Ideally it is always intrinisified.
+			//
+			object result;
```
Comment (jaykrell) [inline mcs/class/corlib/System.Threading/Interlocked.cs]: This probably has to be initialized, or `out`.

File: mono/metadata/icall-def.h [hunk 2]
```diff
@@ -982,7 +982,7 @@ NOHANDLES(ICALL(ILOCK_5, "CompareExchange(int&,int,int)", ves_icall_System_Threa
 NOHANDLES(ICALL(ILOCK_6, "CompareExchange(int&,int,int,bool&)", ves_icall_System_Threading_Interlocked_CompareExchange_Int_Success))
 NOHANDLES(ICALL(ILOCK_7, "CompareExchange(intptr&,intptr,intptr)", ves_icall_System_Threading_Interlocked_CompareExchange_IntPtr))
 NOHANDLES(ICALL(ILOCK_8, "CompareExchange(long&,long,long)", ves_icall_System_Threading_Interlocked_CompareExchange_Long))
-ICALL(ILOCK_9, "CompareExchange(object&,object,object)", ves_icall_System_Threading_Interlocked_CompareExchange_Object)
+NOHANDLES(ICALL(ILOCK_9, "CompareExchange_object_icall", ves_icall_System_Threading_Interlocked_CompareExchange_Object))
```
Comment (jaykrell) [inline mono/metadata/icall-def.h]: `_icall` suffix can be removed.

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

File: mcs/class/corlib/System.Threading/Interlocked.cs (hunk 1 above)
Comment (jaykrell) [reply mcs/class/corlib/System.Threading/Interlocked.cs]: I know it is less efficient, but I went with initialized.
I get nervous about `out` parameters, esp. with all the wrappers. `ref` is very safe.

File: mono/metadata/icall-def.h (hunk 2 above)
Comment (jaykrell) [reply mono/metadata/icall-def.h]: Removed.

**Comment by jaykrell** [comment]
Had to reorder the icalls apparently.

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
Still crashing in tarball build. :(

**Comment by jaykrell** [comment]
Obviously.

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
Possibly superceded by https://github.com/mono/mono/pull/16853 but we can spread the bet.

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build build

**Comment by jaykrell** [comment]
Closing. Superceded by https://github.com/mono/mono/pull/16873.

**Pull request closed without merging**